includes/oauth/class-mailchimp-api.php

```diff
@@ -0,0 +1,177 @@
+<?php
+/**
+ * Newspack's Mailchimp API handeling.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Mailchimp API Settings.
+ */
+class Mailchimp_API {
+	/** API Endpoint Placeholder, needs to replace <dc> by the right datacenter. */
+	const API_ENDPOINT_PLACEHOLDER = 'https://<dc>.api.mailchimp.com/3.0';
+	
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		add_action( 'rest_api_init', [ __CLASS__, 'register_api_endpoints' ] );
+	}
+
+	/**
+	 * Register the endpoints.
+	 */
+	public static function register_api_endpoints() {
+		// Get Mailchimp auth status.
+		\register_rest_route(
+			NEWSPACK_API_NAMESPACE,
+			'/oauth/mailchimp',
+			[
+				'methods'             => \WP_REST_Server::READABLE,
+				'callback'            => [ __CLASS__, 'api_mailchimp_auth_status' ],
+				'permission_callback' => function() {
+					return current_user_can( 'manage_options' );
+				},
+			]
+		);
+
+		// Verify and save Mailchimp API Key.
+		\register_rest_route(
+			NEWSPACK_API_NAMESPACE,
+			'/oauth/mailchimp',
+			[
+				'methods'             => \WP_REST_Server::EDITABLE,
+				'callback'            => [ __CLASS__, 'api_mailchimp_save_key' ],
+				'permission_callback' => function() {
+					return current_user_can( 'manage_options' );
+				},
+			]
+		);
+
+		// Delete Mailchimp API Key.
+		\register_rest_route(
+			NEWSPACK_API_NAMESPACE,
+			'/oauth/mailchimp',
+			[
+				'methods'             => \WP_REST_Server::DELETABLE,
+				'callback'            => [ __CLASS__, 'api_mailchimp_delete_key' ],
+				'permission_callback' => function() {
+					return current_user_can( 'manage_options' );
+				},
+			]
+		);
+	}
+
+	/**
+	 * Check whether we have an API key and its validity.
+	 *
+	 * @return WP_REST_Response
+	 */
+	public static function api_mailchimp_auth_status() {
+		$mailchimp_api_key = get_option( 'newspack_newsletters_mailchimp_api_key' );
```

Comment:
This plugin should define its own option name, and use Newsletters' option as a fallback (and a migration mechanism). Otherwise, if someone, say, removed their Newsletters' key, they will lose their Newspack Plugin's key – and the other way around.